### ISSUE TITLE
Development Discussion: Should we change how groupings are designated?

### ISSUE THREAD

**Issue by wvoigt** [sumpfork/dominiontabs#166]
Some of this may have been covered by discussions in the past, but I could not find it.  A couple of things prompted this topic:

1. There seems to be a growing number of possibe multi-expansion cards (e.g., Events, Landmarks, Boons, Hexes, States, etc).  And right now, any grouping outside of the normal `--special_card_groups` requires code changes.
2. I noticed as I was entering Nocturne, that we have a number of "expansion Extras" sets that I don't think are used anymore.  Remnants of past implementations.

Also, I think it would be nice to see all the multi-expanion groupings handled in a uniform way, with the designations represented in the json instead of the code.  But before that, I think we should discuss what groupings should be implemented.

Right now, the possible groupings are:

- none - every card is printed out individually
- normal groupings - i.e., `--special_card_groups` Cards are grouped together and placed in the normal listing - usually alphabetical.  So this means that Landmarks are sorted along with all the other normal kingdom cards.
- global groupings - ie., `--exclude_events`, `--exclude_landmarks`, etc.  Cards are grouped at a global level all under an "Extras" divider.  A suggestion would be to simplify this to say `--global_card_groups`

What might be missing, is an option for groupings by expansion, with the cards placed in a "<expansion> Extras" divider.  This would keep them out of the alphabetical listing and potentially easier to find/gather for set-up.  A possible option name would be `--expansion_card_groups`.
So cards in `--expansion_card_groups` would be a subset of cards in  `--special_card_groups`, and cards in `--global_card_groups` would be a subset of cards in `--expansion_card_groups`.

So the question is, does this seem worthwhile?

And in any case, I think there should be a discussion on how to represent the global groupings in the json.  A brute force method might be to add properties such as
```
    "card_tag": "Miserable - Twice Miserable",
    "cardset_tags": [
        "nocturne"
    ],
    "cost": "",
    "count": "6",
    "group_tag": "nocturne states",
    "group_extra": true,   <-- New
    "group_global": true,  <-- New
    "types": [
        "State"
    ]    
```

The only issue I see is every card would need this when we really only need it once.  Another option would be to add it only to the card that has the `"group_top": true,` designation (which is used to designate the card that is used for cost and type information for the group).  Another option would be to create some other form of json structure to take care of groups.
Thoughts?

**Comment by nickv2002** [comment]
I agree with your thought process and intent here. I don't want to get into a massive update that delays the rollout of the dividers after Nocturne is fully revealed though.

Couldn't `group_extra` and `group_global` default to _False_?  Then we could just add the tags to be true where they were needed?

**Comment by ossiangrr** [comment]
What I'd like to see is the ability to print the names of associated non-Supply cards right on the tab. So even if they are nowhere near each other in the box, you can see at a glance that, i.e. the Exorcist requires you to go and get the Ghost, Imp, and Will-o'-Wisp decks.
This lets you put those cards in their own space when they go along with multiple kingdom cards.

**Comment by wvoigt** [comment]
So I started to look at this again.  It is not as bad nor as hard as I first thought with a small change.  It even cuts down on the number of sweeps through all the cards.

I noticed that all the cards that we would want to move are always of a specific type.  landmark, event, hex, etc.  So the better place for the information would seem to be with the type information.
I propose the following as an example:

```
    "card_type": [
        "Event"
    ],
    "card_type_image": "event.png",
	"group_extra": true,
	"group_global": true,
	"group_global_card_tag": "events",
	"group_cost": "*",
    "defaultCardCount": 1,
    "tabCostHeightOffset": -1,
    "tabTextHeightOffset": 0
}
```
With `--expansion_card_groups`, `group_extra` when true, means it uses the same card `group_tag` as in `--special_card_groups` but changes the set to be `cardset_tag` += ' extras'.

With `--global_card_groups`, `group_global` can't use the `cardset_tag` because they are different for each set, and these cards need to go across all sets.  There is a need for a way to specify what that grouping is to be called.  Hense the `group_global_card_tag`.  And in this case, the set will be `cardset_tag` = 'extras'.

So I've prototyped this and it seems to be working well.

In working this I started to wonder if it made sense to have a `--kingdom-card-groups` option.  Basically, move any non-Kindgom card (i.e.,  `randomizer` is false) to the expansion extras section.  This is similar, but slightly different and can pull in other cards.  Most are good and make sense (for example, all the Spirit cards in Nocturne).

This produces some correct, but probably odd results.  For example with `--base-cards-with-expansion`, for Alchemy, the only card pulled out is Potion.  For Prosperity, the only ones pulled out are Platinum and Colony.   Most probably do not use `--base-cards-with-expansion`.
**Question**: If `--base-cards-with-expansion` and `--kingdom-card-groups` are both specified, should we do the logical thing and apply both, or should we favor the `--base-cards-with-expansion` and not move the base cards?  Note that for sets like Dominion, doing the logical thing has almost the same effect as running without `--base-cards-with-expansion`, since all the base cards are moved to the "Dominion Extras" section.
